zeppelin-zengine/src/main/java/org/apache/zeppelin/storage/LocalConfigStorage.java

```diff
@@ -95,17 +99,29 @@ public String loadCredentials() throws IOException {
   @Override
   public void saveCredentials(String credentials) throws IOException {
     LOGGER.info("Save Credentials to file: " + credentialPath);
-    writeToFile(credentials, credentialPath);
+    atomicWriteToFile(credentials, credentialPath);
   }
 
   private String readFromFile(File file) throws IOException {
     return IOUtils.toString(new FileInputStream(file));
   }
 
-  private void writeToFile(String content, File file) throws IOException {
-    FileOutputStream out = new FileOutputStream(file);
-    IOUtils.write(content, out);
+  private void atomicWriteToFile(String content, File file) throws IOException {
+    File directory = file.getParentFile();
+    String fileNameParts[] = file.getName().split("\\.");
+    File tempFile = File.createTempFile(fileNameParts[0], fileNameParts[1], directory);
+    FileOutputStream out = new FileOutputStream(tempFile);
+    try {
+      IOUtils.write(content, out);
+    } catch (IOException iox) {
+      tempFile.delete();
+      throw iox;
+    }
     out.close();
+    FileSystem defaultFileSystem = FileSystems.getDefault();
+    Path tempFilePath = defaultFileSystem.getPath(tempFile.getCanonicalPath());
+    Path destinationFilePath = defaultFileSystem.getPath(file.getCanonicalPath());
+    Files.move(tempFilePath, destinationFilePath, StandardCopyOption.ATOMIC_MOVE);
```

Comment:
in my earlier question - what if move throws? should it clean up the tempFilePath?

Reply:
I've added a try-catch here.